packages/autocertifier/src/CertificateCreator.ts

```diff
@@ -0,0 +1,113 @@
+import os from 'os'
+
+const logger = new Logger(module)
+
+export class CertificateCreator {
+
+    private accountPrivateKey?: Buffer
+    private accountPrivateKeyPath: string
+    private readonly acmeDirectoryUrl: string
+    private readonly hmacKid: string
+    private readonly hmacKey: string
+    private readonly challengeInterface: ChallengeInterface
+
+    constructor(acmeDirectoryUrl: string,
+        hmacKid: string,
+        hmacKey: string,
+        privateKeyPath: string,
+        challengeInterface: ChallengeInterface
+    ) {
+        this.acmeDirectoryUrl = acmeDirectoryUrl
+        this.hmacKid = hmacKid
+        this.hmacKey = hmacKey
+        this.challengeInterface = challengeInterface
+
+        if (privateKeyPath.startsWith('~/')) {
+            this.accountPrivateKeyPath = privateKeyPath.replace('~', os.homedir())
+        } else {
+            this.accountPrivateKeyPath = privateKeyPath
+        }
+    }
+
+    public async createCertificate(fqdn: string): Promise<Certificate> {
+
+        logger.info(`Creating certificate for ${fqdn}`)
+        logger.info('Creating acme client')
+
+        const wasNewKeyCreated = await this.createPrivateKey()
+
+        const clientOptions: acme.ClientOptions = {
+            directoryUrl: this.acmeDirectoryUrl,
+            accountKey: this.accountPrivateKey!
+        }
+
+        if (wasNewKeyCreated) {
+            clientOptions.externalAccountBinding = {
+                kid: this.hmacKid,
+                hmacKey: this.hmacKey
+            }
+        }
+
+        const client = new acme.Client(clientOptions)
+
+        logger.info('Creating CSR')
+        const [key, csr] = await acme.crypto.createCsr({
+            commonName: fqdn
+        })
+
+        logger.info('Creating certificate using client.auto')
+        let cert: string
+
+        try {
+            cert = await client.auto({
+                csr,
+                email: 'autocertifier@streamr.network',
+                termsOfServiceAgreed: true,
+                challengePriority: ['dns-01'],
+                challengeCreateFn: async (authz, _challenge, keyAuthorization) => {
+                    await this.challengeInterface.createChallenge(authz.identifier.value,
+                        keyAuthorization)
+                },
+                challengeRemoveFn: async (authz, _challenge, _keyAuthorization) => {
+                    await this.challengeInterface.deleteChallenge(authz.identifier.value)
+                },
+            })
+        } catch (e) {
+            logger.error('Failed to create certificate: ' + e.message)
+            throw e
+        }
+
+        logger.info(`CSR:\n${csr.toString()}`)
+        logger.info(`Private key:\n${key.toString()}`)
+        logger.info(`Certificate:\n${cert.toString()}`)
+
+        return { cert: cert.toString(), key: key.toString() }
+    }
+
+    private createPrivateKey = async (): Promise<boolean> => {
```

Comment:
Would it make sense just to reject if there is an error? (i.e. to change the return type to `Promise<void>`)

Reply:
Added TODO comment